[Known issue][Detection Engine][8.16.0 & 8.16.1] Exception item comment validation on newline chars (#6280)
elastic/security-docs: docs/release-notes/8.16.asciidoc

[discrete]
.The **Exceptions** tab won't properly load if exceptions contain comments with newline characters (`\n`)  
[%collapsible]
====
*Details* +
On December 5, 2024, it was discovered that the **Exceptions** tab will not load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later. 

*Workaround* + 

For custom rules:

. From the {security-app}, <<import-export-rules-ui,export>> the rule or rules with the affected exception lists. 
. Modify the `.ndjson` file so `comments` no longer contain newline characters.
. Return to the {security-app} and <<import-export-rules-ui,re-import>> the rules. Make sure to select the **Overwrite existing exception lists with conflicting "list_id"** option.

For prebuilt rules: 

. Fetch the affected exception list ID or IDs that are associated with the rule. 
.. Find the affected rule's ID (`id`). From the {security-app}, open the rule's details page, go to the page URL, and copy the string at the end. For example, the URL http://host.name/app/security/rules/id/167a5f6f-2148-4792-8226-b5e7a58ef46e contains the `id` `167a5f6f-2148-4792-8226-b5e7a58ef46e` at the end.

> .. Use the `id` to fetch the rule's details using the {api-kibana}/operation/operation-readrule[Retrieve a detection rule API]. Here is an example request that includes the rule ID:

Author: @natasha-moore-elastic I could use your help here 😅 I'm fairly certain that the attribute is correct, but I don't know how to form the second half of the reference. Here are the ESS API docs that I need to reference: https://www.elastic.co/docs/api/doc/kibana/v8/operation/operation-readrule

```suggestion
.. Use the `id` to fetch the rule's details using the {api-kibana}/operation/operation-readrule[Retrieve a detection rule API]. Here is an example request that includes the rule ID:
```

I'll need to do this again for the Serverless changelog [here](https://github.com/elastic/docs-content/pull/233/files#diff-e0d59f20eb51b705c21eb9969115968e0f96c1815e4107c1f237cf2cbbe7aa4cR36). For those release notes, I'll need to reference [these](https://www.elastic.co/docs/api/doc/serverless/operation/operation-readrule) Serverless API docs.

Reviewer: Hey @nastasha-solomon, your reference to the ESS Retrieve a detection rule API looks good to me and works as expected in the preview! 🚀 

For serverless Kibana API docs, I can't see an existing attribute anywhere in the `docs` repo, so I opened a [PR](https://github.com/elastic/docs/pull/3137) to add one. When this is merged, you can use similar syntax to refer to the serverless Retrieve a detection rule API:

`{api-kibana-serverless}/operation/operation-readrule[Retrieve a detection rule API]`

[[known-issue-8.16.1]]
==== Known issues

// tag::known-issue[201820]
[discrete]
.The **Exceptions** tab won't properly load if exceptions contain comments with newline characters (`\n`)  
[%collapsible]
====
*Details* +

> On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later. 

Author: ```suggestion
On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later ({kibana-issue}[#201820]).
```

> On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later ({kibana-issue}[#201820]).

Author: ```suggestion
On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later ({kibana-issue}201820[#201820]).
```

*Workaround* + 

For custom rules:

. From the **Rules** page, <<import-export-rules-ui,export>> the rule or rules with the affected exception lists. 
. Modify the `.ndjson` file so `comments` no longer contain newline characters.
. Return to the **Rules** page and <<import-export-rules-ui,re-import>> the rules. Make sure to select the **Overwrite existing exception lists with conflicting "list_id"** option.

For prebuilt rules: 

> NOTE: If you only need to fix Endpoint exceptions for the Elastic Endpoint rule, follow the above instructions for fixing custom rule exceptions.

Reviewer 2: ```suggestion
NOTE: If you only need to fix exceptions for the Elastic Endpoint rule, you can export and re-import its exception list from the Shared Exception Lists page.
```

@nastasha-solomon this is more accurate, I think, as you won't be able to follow the instructions for custom rules. The Endpoint rule is unique in that it's prebuilt (so can't be exported), but its exception list has special treatment on that Shared Exception Lists page that allows it to be exported (unlike other prebuilt rules).

.. Specify the `id` when fetching the rule's details using the {api-kibana}/operation/operation-readrule[Retrieve a detection rule API]. Here is an example request that includes the `id`:
+
[source,console]
----
curl -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' http://localhost:5601/api/detection_engine/rules?id=167a5f6f-2148-4792-8226-b5e7a58ef46e -u elastic:changeme
----
+
.. The JSON response contains the `id`, `list_id`, and `namespace_type` values within the `exceptions_list` key (as shown below). You need these values when using the Exception list API to retrieve the affected exception list. 
+
[source,console]
----
{
  "id": "167a5f6f-2148-4792-8226-b5e7a58ef46e",
  "exceptions_list": [
    {
      "id": "490525a2-eb66-4320-95b5-88bdd1302dc4",
      "list_id": "f75aae6f-0229-413f-881d-81cb3abfbe2d",
      "namespace_type": "single"
    }
  ]
}
----
+
. Use the export exceptions API to retrieve the affected exception list. Insert the values for the `id`, `list_id`, and `namespace_type` parameters into the following API call:
+
[source,console]
----

> curl -XPOST -u elastic:changeme -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' 'http://localhost:5601/api/exception_lists/_export?list_id=f75aae6f-0229-413f-881d-81cb3abfbe2d&id=490525a2-eb66-4320-95b5-88bdd1302dc4&namespace_type=single' -o list.ndjson

Reviewer 2: ```suggestion
curl -XPOST -H 'Authorization: ApiKey API_KEY_HERE' -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' 'KIBANA_URL/api/exception_lists/_export?list_id=f75aae6f-0229-413f-881d-81cb3abfbe2d&id=490525a2-eb66-4320-95b5-88bdd1302dc4&namespace_type=single' -o list.ndjson
```

Author: Oops, I thought this was only for Serverless. Fix looks good - thanks!

> curl -XPOST -H 'Authorization: ApiKey API_KEY_HERE' -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' 'KIBANA_URL/api/exception_lists/_export?list_id=f75aae6f-0229-413f-881d-81cb3abfbe2d&id=490525a2-eb66-4320-95b5-88bdd1302dc4&namespace_type=single' -o list.ndjson

Reviewer 3: Not sure how we usually document curl requests, do we need to specify that they need to also include the space? If not, ignore.

```suggestion
curl -XPOST -H 'Authorization: ApiKey API_KEY_HERE' -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' '${KIBANA_URL}${SPACE_URL}/api/exception_lists/_export?list_id=f75aae6f-0229-413f-881d-81cb3abfbe2d&id=490525a2-eb66-4320-95b5-88bdd1302dc4&namespace_type=single' -o list.ndjson
```

Author: Good question, I'm not sure. Do users still need to specify the space if they're using the default one?

Reviewer 3: I don't believe so, think just if they need to hit a different space. Might be worth just noting somewhere though to make sure to include space if other than default?

Author: Partially addressed in the 8.16.2 release notes (https://github.com/elastic/security-docs/pull/6290/commits/b348cd66634baf747d3ce6b71fdb300b7e253607).

----
+
. Modify the exception list's `.ndjson` file to ensure `comments[].comment` values don't contain newline characters (`\n`).

> . Re-import the modified exception list using **Import exception lists** option on the **Shared Exception Lists** page (find the page in the navigation menu or by using the {kibana-ref}/introduction.html#kibana-navigation-search[global search field]).

Reviewer 2: I think this reference to `Shared Exception Lists` could use the same link as above, instead of referencing global search.

Author: ```suggestion
. Re-import the modified exception list using **Import exception lists** option on the **<<shared-exception-lists,Shared Exception Lists>>** page.
```

+

> Note that the import will initially fail because the exception list already exists. After the initial failure, an option to overwrite the existing list will appear. Select the option, then resubmit the request to import the corrected exception list.

Reviewer 2: Should this be an explicit `NOTE: ` callout?
```suggestion
NOTE:  The import will initially fail because the exception list already exists. After the initial failure, an option to overwrite the existing list will appear. Select the option, then resubmit the request to import the corrected exception list.
```

Author: Good question - we typically reserve `Note` admonitions for information that's relevant but not critical. Since these details are both relevant and important, I can reword the first sentence so it doesn't come across as an afterthought.

[[known-issue-8.16.0]]
==== Known issues

// tag::known-issue[201820]
[discrete]
.The **Exceptions** tab won't properly load if exceptions contain comments with newline characters (`\n`)  
[%collapsible]
====
*Details* +

> On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later. 

Author: ```suggestion
On December 5, 2024, it was discovered that the **Exceptions** tab won't load properly if any exceptions contain comments with newline characters (`\n`). This issue occurs when you upgrade to 8.16.0 or later ({kibana-issue}[#201820]). 
```

. From the **Rules** page, <<import-export-rules-ui,export>> the rule or rules with the affected exception lists. 
. Modify the `.ndjson` file so `comments` no longer contain newline characters.
. Return to the **Rules** page and <<import-export-rules-ui,re-import>> the rules. Make sure to select the **Overwrite existing exception lists with conflicting "list_id"** option.

For prebuilt rules: 

NOTE: If you only need to fix Endpoint exceptions for the Elastic Endpoint rule, follow the above instructions for fixing custom rule exceptions.

. Follow these steps to fetch the affected exception list ID or IDs that are associated with the rule: 
.. Find the affected rule's ID (`id`). From the **Rules** page, open the details of a rule, go to the page URL, and copy the string at the end. For example, in the URL http://host.name/app/security/rules/id/167a5f6f-2148-4792-8226-b5e7a58ef46e, the string at the end (`167a5f6f-2148-4792-8226-b5e7a58ef46e`) is the `id`.
.. Specify the `id` when fetching the rule's details using the {api-kibana}/operation/operation-readrule[Retrieve a detection rule API]. Here is an example request that includes the `id`:
+
[source,console]
----

> curl -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' http://localhost:5601/api/detection_engine/rules?id=167a5f6f-2148-4792-8226-b5e7a58ef46e -u elastic:changeme

Reviewer 2: ```suggestion
curl -H 'Authorization: ApiKey API_KEY_HERE' -H 'kbn-xsrf: true' -H 'elastic-api-version: 2023-10-31' KIBANA_URL/api/detection_engine/rules?id=167a5f6f-2148-4792-8226-b5e7a58ef46e
```
